### PR TITLE
Simplify/add pyparsing error messages on mathtext/fontconfig errors.

### PR DESCRIPTION
**Pull request by anntzer**
`ParseBaseException.explain(0)` sets up the same "line and caret and
error message" as we previously did manually in _mathtext.  It requires
pyparsing ~2.3.1 (2019)~ 3.0.0 (2022), which is now permissible.

Also use the same detailed explanation on fontconfig pattern parse
failures.

Also suppress expression chaining in these cases, as the explain()
string already contains everything relevant.

(This can easily wait for 3.7 if we don't want to bump the dependency that late in the release cycle.)

-----

(I initially believed we could require pyparsing>=2.3.1, but there was an API change around explain() in 3.0.0 so it seems easier just to require that.)

## PR Checklist

**Tests and Styling**
- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [ ] New features are documented, with examples if plot related.
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

**Comment by anntzer** [comment]
This should now be ready to go.

File: lib/matplotlib/_fontconfig_pattern.py [hunk 1]
```diff
@@ -125,14 +125,10 @@ def parse(self, pattern):
         props = self._properties = {}
         try:
             self._parser.parseString(pattern)
-        except self.ParseException as e:
-            raise ValueError(
-                "Could not parse font string: '%s'\n%s" % (pattern, e)) from e
-
+        except self.ParseException as err:
+            raise ValueError("\n" + err.explain(0)) from None
```
Comment (timhoffm) [inline lib/matplotlib/_fontconfig_pattern.py:129]: I can't judge how much trouble bumping the pyparsing dependency might cause. If we want to be defensive, we could do 

```suggestion
            # can be simplified to err.explain(0) when pyparsing>=3
            message = ParseException.explain(err, 0)
            raise ValueError("\n" + message) from None
```

which works for both versions pyparsing 2 and 3.

Comment (anntzer) [reply lib/matplotlib/_fontconfig_pattern.py]: Sure, went back to requiring 2.3.1.

**Review by timhoffm — COMMENTED** [review]
Should the catch be on ParseException or ParseBaseException? The two places use either one.

**Comment by anntzer** [comment]
I believe this is correct(?), because mathtext sometimes raises a ParseFatalException to stop parsing (and thus needs ParseBaseException) whereas fontconfig_parser doesn't.

**Review by timhoffm — APPROVED** [review]

**Comment by timhoffm** [comment]
Sorry, I accidentally merged with just one review. If anybody feels the PR is not correct please roll back.

**Pull request merged**